### PR TITLE
Migrate data quality guide

### PR DESCRIPTION
**Pull request by jwitz**

### CODE REVIEW

File: concepts/data-quality.md [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+To get the most out of this guide, you should have knowledge of:
+
+- What Airflow is and when to use it. See [Introduction to Apache Airflow](https://www.astronomer.io/guides/intro-to-airflow).
+- Airflow operators. See [Operators 101](https://www.astronomer.io/guides/what-is-an-operator).
+- Airflow connections. See [Managing your Connections in Apache Airflow](https://www.astronomer.io/guides/connections/).
+- Relational databases. See [Relational database on Wikipedia](https://en.wikipedia.org/wiki/Relational_database).
+
+## Data quality essentials
+
+What is considered "good" in terms of data quality varies based on your use case. Defining quality criteria for a given collection of datasets is often a task requiring collaboration with all data professionals involved in each step of the pipeline.
```
Comment (kwcanuck) [inline concepts/data-quality.md:32]:
```suggestion
What is considered good quality data is determined by the needs of your organization. Defining quality criteria for a given collection of datasets is often a task requiring collaboration with all data professionals involved in each step of the pipeline.
```

File: concepts/data-quality.md [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+
+Generally speaking, though, the data quality check process can be summarized in these steps:
```
Comment (kwcanuck) [inline concepts/data-quality.md:34]:
```suggestion
The following is the typical data quality check process:
```

File: concepts/data-quality.md [hunk 3]
```diff
@@ -0,0 +1,456 @@
...
+
+1. Assess where and in what formats relevant data is stored, and where it moves along your pipeline. This step can be assisted by using a data lineage tool.
```
Comment (kwcanuck) [inline concepts/data-quality.md:36]:
```suggestion
- Assess where and in what formats relevant data is stored, and where it moves along your pipeline. A data lineage tool can assist with this process.
```

File: concepts/data-quality.md [hunk 4]
```diff
@@ -0,0 +1,456 @@
...
+2. Gather the data quality criteria from data professionals using the data.
```
Comment (kwcanuck) [inline concepts/data-quality.md:37]:
```suggestion
- Gather the data quality criteria from data professionals using the data.
```

File: concepts/data-quality.md [hunk 5]
```diff
@@ -0,0 +1,456 @@
...
+3. Determine the quality checks to run different components of the dataset.
```
Comment (kwcanuck) [inline concepts/data-quality.md:38]:
```suggestion
- Determine the quality checks to run different components of the dataset.
```

File: concepts/data-quality.md [hunk 6]
```diff
@@ -0,0 +1,456 @@
...
+4. Determine where those quality checks should occur in your pipeline.
```
Comment (kwcanuck) [inline concepts/data-quality.md:39]:
```suggestion
- Determine where those quality checks should occur in your pipeline.
```

File: concepts/data-quality.md [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+5. Determine what happens when a check fails.
+6. Choose a data quality tool (see the "Data quality tools" section of this guide for more information).
```
Comment (kwcanuck) [inline concepts/data-quality.md:41]:
```suggestion
- Choose a data quality tool.
```

File: concepts/data-quality.md [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+7. Write the checks.
```
Comment (kwcanuck) [inline concepts/data-quality.md:42]:
```suggestion
- Write the checks.
```

File: concepts/data-quality.md [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+8. Test the checks.
```
Comment (kwcanuck) [inline concepts/data-quality.md:43]:
```suggestion
- Test the checks.
```

File: concepts/data-quality.md [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+### Types of data quality checks
+
+Data quality checks can be run on different components of the dataset:
+
+- **Column**: Common checks include confirming uniqueness of a key column, restricting the amount of `NULL` values that are allowed, defining logical bounds for numeric or date columns, and defining valid options for categorical variables.
+- **Row**: Common checks include comparing the values in several columns for each row, and ensuring that there are no mutually exclusive values present in the same row.
+- **Table**: Common checks include checking against a template schema to verify that it has not changed, and checking that a table has a minimum row count.
+- **Over several tables**: Common checks include data parity checks across several tables. You can also create logical checks like making sure a customer marked as `inactive` has no purchases listed in any of the other tables.
+
+It is also important to distinguish between the two types of control in data quality:
```
Comment (kwcanuck) [inline concepts/data-quality.md:54]:
```suggestion
The following are the two types of data quality control:
```

File: concepts/data-quality.md [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+
+- **Static control**: Runs checks on Spark or Pandas dataframes, or on existing tables in a database. These checks can be performed anywhere in a typical ETL pipeline and might halt the pipeline in case of a failed check to prevent further processing of data with quality issues.
+- **Flow control**: Monitors incoming data on a record-by-record basis, which sometimes prevents single records that fail certain data quality checks from entering the database. This guide does not cover flow controls.
+
+### Deciding where to place data quality checks
```
Comment (kwcanuck) [inline concepts/data-quality.md:59]:
```suggestion
### Where to run data quality checks
```

File: concepts/data-quality.md [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+
+Data quality checks can be run at different times within a data pipeline or Airflow environment. It often makes sense to test them in their own DAG and later incorporate them into your pipelines to be able to make downstream behavior dependent on the outcome of selected data quality checks.
```
Comment (kwcanuck) [inline concepts/data-quality.md:61]:
```suggestion
Data quality checks can be run at different times within a data pipeline or an Airflow environment. It often makes sense to test them in a dedicated DAG before you incorporate them into your pipelines to make downstream behavior dependent on the outcome of selected data quality checks.
```

File: concepts/data-quality.md [hunk 13]
```diff
@@ -0,0 +1,456 @@
...
+
+For example, within an ETL pipeline, data quality checks can be placed in several different locations as shown in the DAG graph below:
```
Comment (kwcanuck) [inline concepts/data-quality.md:63]:
```suggestion
For example, within an ETL pipeline, data quality checks can be placed in several different locations as shown in the following DAG graph:
```

File: concepts/data-quality.md [hunk 14]
```diff
@@ -0,0 +1,456 @@
...
+
```
Comment (kwcanuck) [inline concepts/data-quality.md:64]: The following DAG graph shows the typical locations for data quality checks:

File: concepts/data-quality.md [hunk 15]
```diff
@@ -0,0 +1,456 @@
...
+- Before the transform step (`data_quality_check_1`)
+- After the transform step (`data_quality_check_2`)
+- Branching off from a step (`data_quality_check_3`)
+
+![Different locations for data quality checks in an ETL pipeline](/img/guides/dq_checks_locations_example_graph.png)
+
+It's common to define further downstream tasks or [`callbacks`](https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/callbacks.html), depending on the results of data quality checks (`post_check_action_1` and `post_check_action_2`). Most commonly these tasks will alert data professionals with [error notifications](https://www.astronomer.io/guides/error-notifications-in-airflow/), such as an email or a Slack message, of a data quality issue. It is also possible to create a downstream task that only runs in case all data quality checks have been successful, for example to log the successful checking event for reporting purposes.
```
Comment (kwcanuck) [inline concepts/data-quality.md:71]:
```suggestion
It's common to use the results of data quality checks (`post_check_action_1` and `post_check_action_2`) to define further downstream tasks or [`callbacks`](https://airflow.apache.org/docs/apache-airflow/stable/logging-monitoring/callbacks.html). Typically, these tasks alert data professionals with [error notifications](https://www.astronomer.io/guides/error-notifications-in-airflow/), such as an email or a Slack message, of a data quality issue. It is also possible to create a downstream task that only runs when all data quality checks are successful. For example, the successful checking event for reporting purposes.
```

File: concepts/data-quality.md [hunk 16]
```diff
@@ -0,0 +1,456 @@
...
+
+When implementing data quality checks, consider how a check success or failure should influence downstream dependencies. [Trigger Rules](https://www.astronomer.io/guides/managing-dependencies/#trigger-rules) are especially useful for managing operator dependencies.
+
+## When to implement data quality checks
+
+When implementing data quality checks, it is important to consider the tradeoffs between the upfront work of implementing checks vs. the cost of downstream issues caused by data with bad quality.
+
+You might need to implement data quality checks if:
```
Comment (kwcanuck) [inline concepts/data-quality.md:79]:
```suggestion
You might need to implement data quality checks in the following circumstances:
```

File: concepts/data-quality.md [hunk 17]
```diff
@@ -0,0 +1,456 @@
...
+
+- Downstream models serve customer-facing results that could be compromised in the case of a data quality issue.
+- There are data quality concerns where an issue might not be immediately obvious to the data professional using the data.
+
+## Data quality tools
+
+There are multiple open source tools that can be used to check data quality from an Airflow DAG. While this guide lists the most commonly used tools here, it focuses on the two tools that also integrate with OpenLineage:
```
Comment (kwcanuck) [inline concepts/data-quality.md:86]:
```suggestion
There are multiple open source tools that can be used to check data quality from an Airflow DAG. This guide focuses on the following tools that integrate with OpenLineage:```

File: concepts/data-quality.md [hunk 18]
```diff
@@ -0,0 +1,456 @@
...
+
+Which tool to choose comes down to individual business needs and preferences. Astronomer recommends using SQL Check operators if you want to:
+
+- Write checks without needing to set up software in addition to Airflow.
+- Write checks as Python dictionaries and in SQL.
+- Use any SQL statement that returns a single row of booleans as a data quality check.
+- Implement many different downstream dependencies depending on the outcome of different checks.
+- Have full observability of which checks failed from within Airflow task logs, including the full SQL statements of failed checks.
+
+Astronomer recommends using a data validation framework like Great Expectations or Soda if:
```
Comment (kwcanuck) [inline concepts/data-quality.md:106]:
```suggestion
Astronomer recommends using a data validation framework such as Great Expectations or Soda in the following circumstances:
```

File: concepts/data-quality.md [hunk 19]
```diff
@@ -0,0 +1,456 @@
...
+- You want to collect the results of your data quality checks in a central place.
+- You prefer to write checks in JSON (Great Expectations) or YAML (Soda) format.
+- Most or all of your checks can be implemented by the predefined checks in the solution of your choice.
+- You want your data quality checks abstracted away from the DAG code.
+
+> **Note**: Currently only SQL Check operators and the `GreatExpectationsOperator` offer data lineage extraction.
+
+### SQL Check operators
+
+> **Note**: You can find more details and examples on SQL Check operators as well as example logs in the [Airflow Data Quality Checks with SQL Operators](https://www.astronomer.io/guides/airflow-sql-data-quality-tutorial/) guide.
```
Comment (kwcanuck) [inline concepts/data-quality.md:117]:
```suggestion
> **Note**: For more details and examples of SQL Check operators and logs, see [Airflow Data Quality Checks with SQL Operators](https://www.astronomer.io/guides/airflow-sql-data-quality-tutorial/).
```

File: concepts/data-quality.md [hunk 20]
```diff
@@ -0,0 +1,456 @@
...
+
+SQL Check operators execute a SQL statement that results in a set of booleans. A result of `True` leads to the check passing and the task being labeled as successful. A result of `False`, or any error when the statement is executed, leads to a failure of the task. Before using any of the operators, you need to define the [connection](https://www.astronomer.io/guides/connections/) to your data storage from the Airflow UI or via an external secrets manager.
```
Comment (kwcanuck) [inline concepts/data-quality.md:119]:
```suggestion
SQL Check operators execute a SQL statement that results in a set of booleans. A result of `True` leads to the check passing and the task being labeled as successful. A result of `False`, or any error when the statement is executed, leads to a failure of the task. Before using any of the operators, you need to define the [connection](https://www.astronomer.io/guides/connections/) to your data storage from the Airflow UI or with an external secrets manager.
```

File: concepts/data-quality.md [hunk 21]
```diff
@@ -0,0 +1,456 @@
...
+- [`SQLCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlcheckoperator): Can be used with any SQL statement that returns a single row of booleans.
+- [`SQLIntervalCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlintervalcheckoperator): Runs checks against historical data.
+- [`SQLValueCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlvaluecheckoperator): Compares the result of a SQL query against a value with or without a tolerance window.
+- [`SQLThresholdCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlthresholdcheckoperator): Compares the result of a SQL query against upper and lower thresholds which may also be described as SQL queries.
+
+The logs from SQL Check operators can be found in the regular Airflow task logs.
+
+### Great Expectations
+
+> **Note**: You can find more information on how to use Great Expectations with Airflow in the [Integrating Airflow and Great Expectations](https://www.astronomer.io/guides/airflow-great-expectations/) guide.
```
Comment (kwcanuck) [inline concepts/data-quality.md:136]:
```suggestion
> **Note**: For more information about using Great Expectations with Airflow, see [Integrating Airflow and Great Expectations](https://www.astronomer.io/guides/airflow-great-expectations/).
```

Comment (kwcanuck) [reply concepts/data-quality.md:136]: Should this be a note?

File: concepts/data-quality.md [hunk 22]
```diff
@@ -0,0 +1,456 @@
...
+
+Great Expectations is an open source data validation framework that allows the user to define data quality checks as a JSON file. The checks, also known as Expectation Suites, can be run in a DAG using the [`GreatExpectationsOperator`](https://registry.astronomer.io/providers/great-expectations/modules/greatexpectationsoperator).
+
+> **Note** All currently available expectations can be viewed on the [Great Expectations website](https://greatexpectations.io/expectations).
```
Comment (kwcanuck) [inline concepts/data-quality.md:140]:
```suggestion
> **Note** All Expectation Suites that are currently available are listed on the [Great Expectations website](https://greatexpectations.io/expectations).
```

File: concepts/data-quality.md [hunk 23]
```diff
@@ -0,0 +1,456 @@
...
+To use Great Expectations, you will need to install the open source Great Expectations package and set up a Great Expectations project with at least one instance of each of the following components:
+
+- Data Context
+- Datasource
+- Expectation Suite
+- Checkpoint
+
+The Great Expectations documentation offers a [step-by-step tutorial](https://docs.greatexpectations.io/docs/tutorials/getting_started/tutorial_overview) for this setup. Additionally, to use Great Expectations with Airflow, you have to install the [Great Expectations provider package](https://registry.astronomer.io/providers/great-expectations).
+
+When using Great Expectations, Airflow task logs show only whether the suite passed or failed. To get a [detailed report](https://docs.greatexpectations.io/docs/terms/data_docs/) on the checks that were run and their results, you can view the HTML files located in the `great_expecations/uncommitted/data_docs/local_site/validations` directory in any browser.
```
Comment (kwcanuck) [inline concepts/data-quality.md:151]:
```suggestion
When using Great Expectations, Airflow task logs show only whether the Expectation Suite passed or failed. To get a [detailed report](https://docs.greatexpectations.io/docs/terms/data_docs/) on the checks that were run and their results, you can view the HTML files located in the `great_expecations/uncommitted/data_docs/local_site/validations` directory in any browser.
```

File: concepts/data-quality.md [hunk 24]
```diff
@@ -0,0 +1,456 @@
...
+
+## Data lineage and data quality
+
+In complex data ecosystems, lineage can be a powerful addition to data quality checks, especially for investigating what data from which origins caused a check to fail.
+
+> **Note**: For more information on data lineage and details on how to set up OpenLineage with Airflow please see the ['OpenLineage and Airflow' guide](https://www.astronomer.io/guides/airflow-openlineage/).
```
Comment (kwcanuck) [inline concepts/data-quality.md:157]:
```suggestion
> **Note**: For more information on data lineage and setting up OpenLineage with Airflow, see the [OpenLineage and Airflow](https://www.astronomer.io/guides/airflow-openlineage/).
```

File: concepts/data-quality.md [hunk 25]
```diff
@@ -0,0 +1,456 @@
...
+- "MY_TEXT_COL" has no null values.
+- "MY_TEXT_COL" has at least 10 distinct values.
+- "MY_NUM_COL" has a minimum value between 90 and 110.
+- `example_table` has at least 1000 rows.
+- The value in each row of "MY_COL_1" plus the value of the same row in "MY_COL_2" is equal to 100.
+- "MY_COL_3" contains only the values `val1`, `val2`, `val3` and `val4`.
+
+### Example: SQL Check operators
+
+The example DAG below consists of 4 tasks:
```
Comment (kwcanuck) [inline concepts/data-quality.md:186]:
```suggestion
The following example DAG consists of four tasks:
```

File: concepts/data-quality.md [hunk 26]
```diff
@@ -0,0 +1,456 @@
...
+
+- First, the`SQLColumnCheckOperator` is used to perform checks on several columns in the target table.
```
Comment (kwcanuck) [inline concepts/data-quality.md:188]:
```suggestion
- The`SQLColumnCheckOperator` is used to perform checks on several columns in the target table.
```

File: concepts/data-quality.md [hunk 27]
```diff
@@ -0,0 +1,456 @@
...
+- Next, two `SQLTableCheckOperators` perform one check each. One operator checks several columns, while the other checks the entire table.
```
Comment (kwcanuck) [inline concepts/data-quality.md:189]:
```suggestion
- Two `SQLTableCheckOperators` perform one check each. One operator checks several columns, while the other checks the entire table.
```

File: concepts/data-quality.md [hunk 28]
```diff
@@ -0,0 +1,456 @@
...
+- Finally, the `SQLCheckOperator` is used to make sure a categorical variable is set to one of 4 options.
+
+While this example shows all the checks being written within the Python file defining the DAG, it is possible to modularize commonly used checks and SQL statements in separate files (for Astro CLI users, those files can be put into the `/include` directory).
```
Comment (kwcanuck) [inline concepts/data-quality.md:192]:
```suggestion
While this example shows all the checks being written within the Python file defining the DAG, it is possible to modularize commonly used checks and SQL statements in separate files. If you're using the Astro CLI, you can add the files to the `/include` directory.
```

File: concepts/data-quality.md [hunk 29]
```diff
@@ -0,0 +1,456 @@
...
+
+    start >> [column_checks, table_checks_aggregated,
+              table_checks_not_aggregated, check_val_in_list] >> end
+```
+
+### Example: Great Expectations
+
+The following example runs the same data quality checks as the SQL Check operators example against the same database. After setting up the Great Expectations instance with all four components, the checks can be defined in a JSON file to form an Expectation Suite.
+
+> **Note**: For each of the checks in this example, an `Expectation` already exists. This is not always the case, and for more complicated checks you may need to define a [custom expectation](https://docs.greatexpectations.io/docs/guides/expectations/creating_custom_expectations/overview).
```
Comment (kwcanuck) [inline concepts/data-quality.md:308]:
```suggestion
> **Note**: For each of the checks in this example, an `Expectation Suite` already exists. This is not always the case, and for more complicated checks you may need to define a [custom Expectation Suite](https://docs.greatexpectations.io/docs/guides/expectations/creating_custom_expectations/overview).
```

**Review by kwcanuck — APPROVED** [review]

File: concepts/data-quality.md [hunk 30]
```diff
@@ -0,0 +1,456 @@
...
+- Have full observability of which checks failed from within Airflow task logs, including the full SQL statements of failed checks.
+
+Astronomer recommends using a data validation framework like Great Expectations or Soda if:
+
+- You want to collect the results of your data quality checks in a central place.
+- You prefer to write checks in JSON (Great Expectations) or YAML (Soda) format.
+- Most or all of your checks can be implemented by the predefined checks in the solution of your choice.
+- You want your data quality checks abstracted away from the DAG code.
+
+> **Note**: Currently only SQL Check operators and the `GreatExpectationsOperator` offer data lineage extraction.
```
Comment (jwitz) [inline concepts/data-quality.md:113]:
```suggestion
Currently only SQL Check operators and the `GreatExpectationsOperator` offer data lineage extraction.
```

File: concepts/data-quality.md [hunk 31]
```diff
@@ -0,0 +1,456 @@
...
+- [`SQLTableCheckOperator`](https://registry.astronomer.io/providers/common-sql/modules/sqltablecheckoperator): Can run aggregated and non-aggregated statements involving several columns of a table.
+- [`SQLCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlcheckoperator): Can be used with any SQL statement that returns a single row of booleans.
+- [`SQLIntervalCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlintervalcheckoperator): Runs checks against historical data.
+- [`SQLValueCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlvaluecheckoperator): Compares the result of a SQL query against a value with or without a tolerance window.
+- [`SQLThresholdCheckOperator`](https://registry.astronomer.io/providers/apache-airflow/modules/sqlthresholdcheckoperator): Compares the result of a SQL query against upper and lower thresholds which may also be described as SQL queries.
+
+The logs from SQL Check operators can be found in the regular Airflow task logs.
```
Comment (jwitz) [inline concepts/data-quality.md:132]:
```suggestion
The logs from SQL Check operators can be found in the regular Airflow task logs. For more details and examples of SQL Check operators and logs, see [Airflow Data Quality Checks with SQL Operators](https://www.astronomer.io/guides/airflow-sql-data-quality-tutorial/).
```

File: concepts/data-quality.md (hunk 22 above)
Comment (jwitz) [inline concepts/data-quality.md:140]:
```suggestion
Great Expectations is an open source data validation framework that allows the user to define data quality checks as a JSON file. The checks, also known as Expectation Suites, can be run in a DAG using the [`GreatExpectationsOperator`](https://registry.astronomer.io/providers/great-expectations/modules/greatexpectationsoperator). All currently available expectations can be viewed on the [Great Expectations website](https://greatexpectations.io/expectations).
```

**Pull request merged**